File: voip_reactjs/src/services/configService.ts
import { AppConfig, DEFAULT_CONFIG } from '../types';

export class ConfigService {
  private static readonly STORAGE_KEY = 'mediaSFUCredentials'; // Same key as react_ref
  private static readonly CONFIG_KEY = 'voipAppConfig'; // New key for full config
  private config: AppConfig;
  private listeners: Array<(config: AppConfig) => void> = [];

  constructor() {
    this.config = this.loadConfig();
  }

  // Load configuration from localStorage (same pattern as react_ref)
  private loadConfig(): AppConfig {
    try {
      // First try to load full config
      const fullConfig = localStorage.getItem(ConfigService.CONFIG_KEY);
      if (fullConfig) {
        const parsed = JSON.parse(fullConfig);
        return { ...DEFAULT_CONFIG, ...parsed };
      }

      // Fallback to old credential format for backward compatibility
      const stored = localStorage.getItem(ConfigService.STORAGE_KEY);
      if (stored) {
        const credentials = JSON.parse(stored);
        return {
          ...DEFAULT_CONFIG,
          api: {
            ...DEFAULT_CONFIG.api,
            key: credentials.apiKey || '',
            userName: credentials.apiUserName || ''
          }
        };
      }
    } catch (error) {
      // Handle config load error silently
    }
    return { ...DEFAULT_CONFIG };
  }

  // Save configuration to localStorage (same pattern as react_ref)
  private saveConfig(): void {
    try {
      // Save full config to new key
      localStorage.setItem(ConfigService.CONFIG_KEY, JSON.stringify(this.config));

      // Also maintain backward compatibility with old format
      const credentials = {
        apiUserName: this.config.api.userName,
        apiKey: this.config.api.key
      };
      
      localStorage.setItem(ConfigService.STORAGE_KEY, JSON.stringify(credentials));
      this.notifyListeners();
    } catch (error) {
      // Handle save error silently
    }
  }

  // Get current configuration
  getConfig(): AppConfig {
    return { ...this.config };
  }

  // Update configuration
  updateConfig(updates: Partial<AppConfig>): void {
    this.config = { ...this.config, ...updates };
    this.saveConfig();
  }

  // Update API configuration
  updateApiConfig(apiConfig: Partial<AppConfig['api']>): void {
    this.config.api = { ...this.config.api, ...apiConfig };
    this.saveConfig();
  }

  // Update realtime configuration
  updateRealtimeConfig(realtimeConfig: Partial<AppConfig['realtime']>): void {
    this.config.realtime = { ...this.config.realtime, ...realtimeConfig };
    this.saveConfig();
  }

  // Update UI configuration
  updateUIConfig(uiConfig: Partial<AppConfig['ui']>): void {
    this.config.ui = { ...this.config.ui, ...uiConfig };
    this.saveConfig();
  }

  // Update calls configuration
  updateCallsConfig(callsConfig: Partial<AppConfig['calls']>): void {
    this.config.calls = { ...this.config.calls, ...callsConfig };
    this.saveConfig();
  }

  // Check if API is configured
  isApiConfigured(): boolean {
    return !!(this.config.api.key && this.config.api.userName && this.config.api.baseUrl);
  }

  // Get API credentials
  getApiCredentials() {
    return {
      apiKey: this.config.api.key,
      apiUserName: this.config.api.userName,
      baseUrl: this.config.api.baseUrl
    };
  }

  // Reset to defaults
  resetConfig(): void {
    this.config = { ...DEFAULT_CONFIG };
    this.saveConfig();
  }

  // Export configuration as JSON
  exportConfig(): string {
    return JSON.stringify(this.config, null, 2);
  }

  // Import configuration from JSON
  importConfig(configJson: string): boolean {
    try {
      const importedConfig = JSON.parse(configJson);
      // Validate the structure
      if (this.validateConfig(importedConfig)) {
        this.config = { ...DEFAULT_CONFIG, ...importedConfig };
        this.saveConfig();
        return true;
      }
      return false;
    } catch (error) {
      // Handle import error silently
      return false;
    }
  }

  // Validate configuration structure
  private validateConfig(config: any): boolean {
    // Basic validation - ensure required properties exist
    return (
      config &&
      typeof config === 'object' &&
      config.api &&
      config.realtime &&
      config.ui &&
      config.calls
    );
  }

  // Subscribe to configuration changes
  subscribe(listener: (config: AppConfig) => void): () => void {
    this.listeners.push(listener);
    // Return unsubscribe function
    return () => {
      const index = this.listeners.indexOf(listener);
      if (index > -1) {
        this.listeners.splice(index, 1);
      }
    };
  }

  // Notify all listeners of configuration changes
  private notifyListeners(): void {
    this.listeners.forEach(listener => {
      try {
        listener(this.getConfig());
      } catch (error) {
        // Handle listener error silently
      }
    });
  }

  // Get theme
  getTheme(): 'light' | 'dark' {
    return this.config.ui.theme;
  }

  // Toggle theme
  toggleTheme(): void {
    this.config.ui.theme = this.config.ui.theme === 'light' ? 'dark' : 'light';
    this.saveConfig();
  }

  // Get realtime settings
  getRealtimeSettings() {
    return {
      enabled: this.config.realtime.enabled,
      interval: this.config.realtime.interval
    };
  }

  // Get call settings
  getCallSettings() {
    return {
      autoAnswer: this.config.calls.autoAnswer,
      recordCalls: this.config.calls.recordCalls,
      defaultRingTime: this.config.calls.defaultRingTime
    };
  }
}

export default ConfigService;
export const configService = new ConfigService();

File: voip_flutter/lib/services/config_service.dart
// Config service - Dart equivalent of TypeScript configService.ts
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../types/config_types.dart';
import '../utils/logger.dart';

class ConfigService {
  static const String _configKey = 'voip_app_config';
  static const String _credentialsKey = 'mediaSFUCredentials';

  /// Load app configuration from SharedPreferences
  Future<AppConfig> loadConfig() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final configJson = prefs.getString(_configKey);

      if (configJson != null) {
        final configMap = jsonDecode(configJson) as Map<String, dynamic>;
        return AppConfig.fromJson(configMap);
      }

      // Return default config if no saved config exists
      return defaultConfig;
    } catch (error) {
      Logger.error('Error loading config: $error');
      return defaultConfig;
    }
  }

  /// Save app configuration to SharedPreferences
  Future<bool> saveConfig(AppConfig config) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final configJson = jsonEncode(config.toJson());
      await prefs.setString(_configKey, configJson);

      // Also save MediaSFU credentials in the expected format
      final credentialsMap = {
        'apiUserName': config.api.userName,
        'apiKey': config.api.key,
      };
      await prefs.setString(_credentialsKey, jsonEncode(credentialsMap));

      // ALSO save individual credentials for MediaSFU handler compatibility
      await prefs.setString('mediasfu_api_username', config.api.userName);
      await prefs.setString('mediasfu_api_key', config.api.key);

      return true;
    } catch (error) {
      Logger.error('Error saving config: $error');
      return false;
    }
  }

  /// Get MediaSFU credentials from configuration
  Future<Map<String, String>?> getCredentials() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final credentialsJson = prefs.getString(_credentialsKey);

      if (credentialsJson != null) {
        final credentials = jsonDecode(credentialsJson) as Map<String, dynamic>;
        final apiUserName = credentials['apiUserName'] as String?;
        final apiKey = credentials['apiKey'] as String?;

        if (apiUserName != null &&
            apiKey != null &&
            apiUserName.isNotEmpty &&
            apiKey.isNotEmpty) {
          return {
            'apiUserName': apiUserName,
            'apiKey': apiKey,
          };
        }
      }
    } catch (error) {
      Logger.error('Error getting credentials: $error');
    }

    return null;
  }

  /// Check if API is properly configured
  Future<bool> isApiConfigured() async {
    final credentials = await getCredentials();
    if (credentials == null) return false;

    final apiUserName = credentials['apiUserName'];
    final apiKey = credentials['apiKey'];

    // Check if credentials meet the validation requirements
    return apiUserName != null &&
        apiKey != null &&
        apiUserName.length >= 6 &&
        apiKey.length == 64 &&
        RegExp(r'^[a-zA-Z0-9]+$').hasMatch(apiUserName) &&
        RegExp(r'^[a-fA-F0-9]{64}$').hasMatch(apiKey);
  }

  /// Update API configuration
  Future<bool> updateApiConfig({
    required String apiKey,
    required String apiUserName,
    String? baseUrl,
  }) async {
    try {
      final currentConfig = await loadConfig();
      final updatedConfig = currentConfig.copyWith(
        api: currentConfig.api.copyWith(
          key: apiKey,
          userName: apiUserName,
          baseUrl: baseUrl ?? currentConfig.api.baseUrl,
        ),
      );

      return await saveConfig(updatedConfig);
    } catch (error) {
      Logger.error('Error updating API config: $error');
      return false;
    }
  }

  /// Update realtime configuration
  Future<bool> updateRealtimeConfig({
    bool? enabled,
    int? interval,
  }) async {
    try {
      final currentConfig = await loadConfig();
      final updatedConfig = currentConfig.copyWith(
        realtime: currentConfig.realtime.copyWith(
          enabled: enabled ?? currentConfig.realtime.enabled,
          interval: interval ?? currentConfig.realtime.interval,
        ),
      );

      return await saveConfig(updatedConfig);
    } catch (error) {
      Logger.error('Error updating realtime config: $error');
      return false;
    }
  }

  /// Update UI configuration
  Future<bool> updateUiConfig({
    AppTheme? theme,
    bool? compactMode,
  }) async {
    try {
      final currentConfig = await loadConfig();
      final updatedConfig = currentConfig.copyWith(
        ui: currentConfig.ui.copyWith(
          theme: theme ?? currentConfig.ui.theme,
          compactMode: compactMode ?? currentConfig.ui.compactMode,
        ),
      );

      return await saveConfig(updatedConfig);
    } catch (error) {
      Logger.error('Error updating UI config: $error');
      return false;
    }
  }

  /// Update calls configuration
  Future<bool> updateCallsConfig({
    bool? autoAnswer,
    bool? recordCalls,
    int? defaultRingTime,
  }) async {
    try {
      final currentConfig = await loadConfig();
      final updatedConfig = currentConfig.copyWith(
        calls: currentConfig.calls.copyWith(
          autoAnswer: autoAnswer ?? currentConfig.calls.autoAnswer,
          recordCalls: recordCalls ?? currentConfig.calls.recordCalls,
          defaultRingTime:
              defaultRingTime ?? currentConfig.calls.defaultRingTime,
        ),
      );

      return await saveConfig(updatedConfig);
    } catch (error) {
      Logger.error('Error updating calls config: $error');
      return false;
    }
  }

  /// Export configuration as JSON string
  Future<String?> exportConfig() async {
    try {
      final config = await loadConfig();
      return jsonEncode(config.toJson());
    } catch (error) {
      Logger.error('Error exporting config: $error');
      return null;
    }
  }

  /// Import configuration from JSON string
  Future<bool> importConfig(String configJson) async {
    try {
      final configMap = jsonDecode(configJson) as Map<String, dynamic>;
      final config = AppConfig.fromJson(configMap);
      return await saveConfig(config);
    } catch (error) {
      Logger.error('Error importing config: $error');
      return false;
    }
  }

  /// Reset configuration to defaults
  Future<bool> resetConfig() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(_configKey);
      await prefs.remove(_credentialsKey);
      return true;
    } catch (error) {
      Logger.error('Error resetting config: $error');
      return false;
    }
  }

  /// Validate API credentials format
  bool validateCredentials(String apiUserName, String apiKey) {
    // API Username validation: alphanumeric, at least 6 characters
    if (apiUserName.length < 6 ||
        !RegExp(r'^[a-zA-Z0-9]+$').hasMatch(apiUserName)) {
      return false;
    }

    // API Key validation: exactly 64 hexadecimal characters
    if (apiKey.length != 64 || !RegExp(r'^[a-fA-F0-9]{64}$').hasMatch(apiKey)) {
      return false;
    }

    return true;
  }

  /// Get current theme
  Future<AppTheme> getCurrentTheme() async {
    final config = await loadConfig();
    return config.ui.theme;
  }

  /// Toggle theme between light and dark
  Future<AppTheme> toggleTheme() async {
    final currentConfig = await loadConfig();
    final newTheme = currentConfig.ui.theme == AppTheme.light
        ? AppTheme.dark
        : AppTheme.light;

    await updateUiConfig(theme: newTheme);
    return newTheme;
  }

  /// Get live updates configuration
  Future<RealtimeConfig> getRealtimeConfig() async {
    final config = await loadConfig();
    return config.realtime;
  }

  /// Get calls configuration
  Future<CallsConfig> getCallsConfig() async {
    final config = await loadConfig();
    return config.calls;
  }
}
